library/core/src/main/java/com/google/android/exoplayer2/metadata/dvbsi/AitDecoder.java

```diff
@@ -0,0 +1,195 @@
+            timestampAdjuster.adjustSampleTimestamp(inputBuffer.timeUs - inputBuffer.subsampleOffsetUs);
+        }
+
+        ByteBuffer buffer = inputBuffer.data;
+        byte[] data = buffer.array();
+        int size = buffer.limit();
+        sectionData.reset(data, size);
+
+        int tableId = sectionData.peekUnsignedByte();
+
+        //Only this table is allowed in AIT streams
+        if (tableId == 0x74) {
+            return parseAit(sectionData);
+        }
+
+        return new Metadata();
+    }
+
+    private Metadata parseAit(ParsableByteArray sectionData) {
+        int tmp;
+
+        int tableId = sectionData.readUnsignedByte();
+
+        tmp = sectionData.readUnsignedShort();
+        int endOfSection = sectionData.getPosition() + (tmp & 4095) - 4 /* Ignore leading CRC */;
+
+        tmp = sectionData.readUnsignedShort();
+        int applicationType = tmp & 0x7fff;
+
+        tmp = sectionData.readUnsignedByte();
+        int versionNumber = (tmp & 0x3e) >> 1;
+        boolean current = (tmp & 1) == 1;
+
+        int section_number = sectionData.readUnsignedByte();
+        int last_section_number = sectionData.readUnsignedByte();
+
+        tmp = sectionData.readUnsignedShort();
+        int commonDescriptorsLength = tmp & 4095;
+
+        //Since we currently only keep url and control code, which are unique per application,
+        //there is no useful information in common descriptor.
+        sectionData.skipBytes(commonDescriptorsLength);
+
+        tmp = sectionData.readUnsignedShort();
+        int appLoopLength = tmp & 4095;
+
+        ArrayList<Ait> aits = new ArrayList<>();
+        while(sectionData.getPosition() < endOfSection) {
+            // Values that will be stored in Ait()
+            String aitUrlBase = null;
+            String aitUrlExtension = null;
+            int aitControlCode = -1;
+
+            long application_identifier = sectionData.readUnsignedInt24() << 24L;
+            application_identifier |= sectionData.readUnsignedInt24();
+            int controlCode = sectionData.readUnsignedByte();
+
+            aitControlCode = controlCode;
+
+            tmp = sectionData.readUnsignedShort();
+            int sectionLength = tmp & 4095;
+            int positionOfNextSection = sectionData.getPosition() + sectionLength;
+            while(sectionData.getPosition() < positionOfNextSection) {
+                int type = sectionData.readUnsignedByte();
+                int l = sectionData.readUnsignedByte();
+                int positionOfNextSection2 = sectionData.getPosition() + l;
+
+                if(type == DESCRIPTOR_TRANSPORT_PROTOCOL) {
+                    int protocolId = sectionData.readUnsignedShort();
+                    int label = sectionData.readUnsignedByte();
+
+                    if(protocolId == TRANSPORT_PROTOCOL_HTTP) {
+                        while (sectionData.getPosition() < positionOfNextSection2) {
```

Comment:
instead of using a `2` suffix, can we call this `positionOfNextSelector` (and also rename `positionOfNextSection` to `positionOfNextDescriptor` above)

Same also for any of the `l` or `len` variables, especially if their scope is quite large, could we qualify them with what they're the length of?

(this is assuming I've correctly matched up your code with the names used in the spec! Please correct me if I'm making nonsense suggestions)

Reply:
Right, that's indeed a mess, I've refactored this.
FWIW I don't end with the same renaming as you.
Hopefully I added enough comments and reference to the standard so we can converge :-)